File: src/ski-coach-agent-python/tools/coach_tools.py
"""
Ski Coach Tools - AI agent tools for slope recommendations and day planning.
"""
import json
from typing import Optional, Annotated

from pydantic import Field
from agent_framework import tool

from services.coach_service import CoachService

_coach_service = CoachService()


@tool(name="recommend_slope", description="Get personalized slope recommendations based on skill level and preferences")
async def recommend_slope(
    skill_level: Annotated[str, Field(description="Skier skill level: 'beginner', 'intermediate', 'advanced', or 'expert'")],
    preferences: Annotated[Optional[str], Field(description="Optional comma-separated preferences like 'avoid_crowds,groomed_only'")] = None,
) -> str:
    prefs_dict = None
    if preferences:
        prefs_dict = {p.strip(): True for p in preferences.lower().split(',') if p.strip()}
    result = await _coach_service.recommend_slope(skill_level, prefs_dict)
    return json.dumps(result, indent=2)


@tool(name="build_day_plan", description="Build a full day ski plan with morning, midday, and afternoon recommendations")
async def build_day_plan(
    skill_level: Annotated[str, Field(description="Skier skill level: 'beginner', 'intermediate', 'advanced', or 'expert'")],
) -> str:
    result = await _coach_service.build_day_plan(skill_level)
    return json.dumps(result, indent=2)

File: src/ski-coach-skill-dotnet/Tools/CoachTools.cs
using System.ComponentModel;
using Microsoft.Extensions.AI;
using ModelContextProtocol.Server;
using SkiCoachSkill.Dotnet.Services;

namespace SkiCoachSkill.Dotnet.Tools;

/// <summary>
/// The ski-coach skill's tools ("scripts"), ported from <c>ski-coach-agent-python</c>'s
/// <c>tools/coach_tools.py</c>. Each method is registered as a literal MCP tool via
/// <see cref="McpServerToolAttribute"/> — the "script" a skills-based orchestrator invokes after discovering
/// this skill over MCP (see <c>Skills/CoachSkillResources.cs</c>).
/// </summary>
[McpServerToolType]
public class CoachTools
{
    private readonly CoachDataService _coachDataService;

    public CoachTools(CoachDataService coachDataService)
    {
        _coachDataService = coachDataService;
    }

    [McpServerTool(Name = "recommend_slope")]
    [Description("Get personalized slope recommendations based on skill level and preferences")]
    public async Task<string> RecommendSlope(
        [Description("Skier skill level: 'beginner', 'intermediate', 'advanced', or 'expert'")] string skillLevel,
        [Description("Optional comma-separated preferences like 'avoid_crowds,groomed_only'")] string? preferences = null)
    {
        return await _coachDataService.RecommendSlopeAsync(skillLevel, preferences);
    }

    [McpServerTool(Name = "build_day_plan")]
    [Description("Build a full day ski plan with morning, midday, and afternoon recommendations")]
    public async Task<string> BuildDayPlan(
        [Description("Skier skill level: 'beginner', 'intermediate', 'advanced', or 'expert'")] string skillLevel)
    {
        return await _coachDataService.BuildDayPlanAsync(skillLevel);
    }

    /// <summary>Used only to derive the SKILL.md "Scripts" table from the same [Description] metadata above.</summary>
    public IEnumerable<AIFunction> GetFunctions()
    {
        return
        [
            AIFunctionFactory.Create(RecommendSlope),
            AIFunctionFactory.Create(BuildDayPlan)
        ];
    }
}
